src/main/java/net/neoforged/neoforge/fluids/crafting/FluidIngredient.java

```diff
@@ -0,0 +1,292 @@
+                    } else if (ingredient instanceof TagFluidIngredient tag) {
+                        return Either.right(tag);
+                    }
+                    throw new IllegalStateException("Basic fluid ingredient should be either a fluid or a tag!");
+                });
+    }
+
+    private static MapCodec<FluidIngredient> makeMapCodec() {
+        return NeoForgeExtraCodecs.<FluidIngredientType<?>, FluidIngredient, FluidIngredient>dispatchMapOrElse(
+                NeoForgeRegistries.FLUID_INGREDIENT_TYPES.byNameCodec(),
+                FluidIngredient::getType,
+                FluidIngredientType::codec,
+                FluidIngredient.SINGLE_OR_TAG_CODEC).xmap(either -> either.map(id -> id, id -> id), ingredient -> {
+                    // prefer serializing without a type field, if possible
+                    if (ingredient instanceof SingleFluidIngredient || ingredient instanceof TagFluidIngredient) {
+                        return Either.right(ingredient);
+                    }
+
+                    return Either.left(ingredient);
+                });
+    }
+
+    private static Codec<FluidIngredient> codec(boolean allowEmpty) {
+        var listCodec = Codec.lazyInitialized(() -> allowEmpty ? LIST_CODEC : LIST_CODEC_NON_EMPTY);
+        return Codec.either(listCodec, MAP_CODEC_CODEC)
+                // [{...}, {...}] is turned into a CompoundIngredient instance
+                .xmap(either -> either.map(CompoundFluidIngredient::of, i -> i),
+                        ingredient -> {
+                            // serialize CompoundIngredient instances as an array over their children
+                            if (ingredient instanceof CompoundFluidIngredient compound) {
+                                return Either.left(compound.children());
+                            } else if (ingredient.isEmpty()) {
+                                // serialize empty ingredients as []
+                                return Either.left(List.of());
+                            }
+
+                            return Either.right(ingredient);
+                        });
+    }
+
+    // empty
+    public static FluidIngredient empty() {
+        return EmptyFluidIngredient.INSTANCE;
+    }
+
+    // convenience methods
+    public static FluidIngredient of() {
+        return empty();
+    }
+
+    public static FluidIngredient of(FluidStack... fluids) {
+        return of(Arrays.stream(fluids).map(FluidStack::getFluid));
+    }
+
+    public static FluidIngredient of(Fluid... fluids) {
+        return of(Arrays.stream(fluids));
+    }
+
+    private static FluidIngredient of(Stream<Fluid> fluids) {
+        return CompoundFluidIngredient.of(fluids.map(FluidIngredient::single));
+    }
+
+    public static FluidIngredient single(FluidStack stack) {
+        return single(stack.getFluid());
+    }
+
+    public static FluidIngredient single(Fluid fluid) {
+        return single(fluid.builtInRegistryHolder());
+    }
```

Comment:
IMO these are redundant and only the vararg methods are enough. But if you want to keep them it's ok.

Reply:
I would like to keep them as they avoid unnecessary creation of CompoundFluidIngredient, though that behaviour could also just be rolled over into the varargs method

Reply:
I mean that `CompoundFluidIngredient.of` should already unwrap?

Reply:
That's true, the stream being constructed and all would still mean *some* overhead though

Reply:
I don't think that instantiating ingredients in code is perf-critical. Could have a shortcut on the vararg method otherwise. :D

Reply:
Yeah that's fair! I'll keep em around for now, see if people actually use them, and if not we can just ditch them